WalletWasabi/WabiSabi/Crypto/UnboundedChannel.cs

```diff
@@ -0,0 +1,37 @@
+using System.Threading;
+using System.Threading.Tasks;
+using System.Collections.Concurrent;
+
+namespace WalletWasabi.WabiSabi.Crypto
+{
+	public class UnboundedChannel<T>
+	{
+		private readonly ConcurrentQueue<T> _queue = new();
+		private readonly ConcurrentQueue<TaskCompletionSource<T>> _waitingQueue = new();
+
+		public void Send(T item)
+		{
+			if (_waitingQueue.TryDequeue(out var tcs))
+			{
+				tcs.TrySetResult(item);
+				return;
+			}
+
+			_queue.Enqueue(item);
+		}
+
+		public Task<T> TakeAsync(CancellationToken cancellationToken = default(CancellationToken))
+		{
+			if (_queue.TryDequeue(out var item))
+			{
+				return Task.FromResult(item);
+			}
+
+			var tcs = new TaskCompletionSource<T>();
```

Comment:
If I call Send here? Maybe locks and lists are better?

Reply:
i would prefer the built in channel type, if there's a good reason not to use it i missed it in my review